agent/lib_monolog.c

```diff
@@ -0,0 +1,387 @@
+
+  nr_php_zend_hash_zval_apply(context,
+                              (nr_php_zval_apply_t)nr_monolog_fmt_context_item,
+                              (void*)&strbuf TSRMLS_CC);
+
+  return nr_str_append(strbuf, "]", "");
+}
+
+/*
+ * Purpose : Convert $context argument of Monolog\Logger::addRecord to a string
+ *
+ * Params  : # of Monolog\Logger::addRecord arguments, and
+ * Monolog\Logger::addRecord argument list
+ *
+ * Returns : A new string with Monolog's log context
+ */
+static char* nr_monolog_get_context(size_t argc, NR_EXECUTE_PROTO TSRMLS_DC) {
+  char* context = nr_strdup("");
+  zval* context_arg = NULL;
+
+  if (3 > argc) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s: $context not available", __func__);
+    goto return_context;
+  }
+
+  context_arg = nr_php_arg_get(3, NR_EXECUTE_ORIG_ARGS TSRMLS_CC);
+
+  if (!nr_php_is_zval_valid_array(context_arg)) {
+    nrl_verbosedebug(NRL_INSTRUMENT,
+                     "%s: expected $context be a valid array, got type %d",
+                     __func__, Z_TYPE_P(context_arg));
+    goto return_context;
+  }
+
+  if (0 == nr_php_zend_hash_num_elements(Z_ARRVAL_P(context_arg))) {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s: $context has no elements", __func__);
+    goto return_context;
+  }
+
+  context = nr_monolog_fmt_context(context, Z_ARRVAL_P(context_arg) TSRMLS_CC);
+
+return_context:
+  nr_php_arg_release(&context_arg);
+  return context;
+}
+
+/*
+ * Purpose : Combine $message and $context arguments of
+ * Monolog\Logger::addRecord into a single string to be used as a message
+ * property of the log event.
+ *
+ * Params  : # of Monolog\Logger::addRecord arguments, and
+ * Monolog\Logger::addRecord argument list
+ *
+ * Returns : A new string with a log record message; caller must free
+ */
+static char* nr_monolog_build_message(size_t argc, NR_EXECUTE_PROTO TSRMLS_DC) {
+  char* message_and_context = nr_strdup("");
+
+  char* message = nr_monolog_get_message(NR_EXECUTE_ORIG_ARGS TSRMLS_CC);
+  message_and_context = nr_str_append(message_and_context, message, "");
+  nr_free(message);
+
+  char* context = nr_monolog_get_context(argc, NR_EXECUTE_ORIG_ARGS TSRMLS_CC);
+  message_and_context = nr_str_append(message_and_context, context, " ");
+  nr_free(context);
+
+  return message_and_context;
+}
+
+/*
+ * Purpose : Create timestamp for the log event by inspecting $datetime argument
+ * of Monolog\Logger::addRecord
+ *
+ * Params  : Logger API, # of Monolog\Logger::addRecord arguments, and
+ * Monolog\Logger::addRecord argument list
+ *
+ * Returns : timestamp in milliseconds calculated from $datetime if available,
+ * current time otherwise.
+ */
+static nrtime_t nr_monolog_get_timestamp(int api,
+                                         size_t argc,
+                                         NR_EXECUTE_PROTO TSRMLS_DC) {
+  nrtime_t timestamp = nr_get_time();
+
+  zval* datetime = NULL;
+  /* $datetime is only available since API level 2 */
+  if (2 <= api && 4 <= argc) {
```

Comment:
Maybe make the 2 and 4 something like MONOLOG_API_2 and MONOLOG_API_4 just to give some meaning to the values?

Reply:
I don't think it will improve code readability. It should be clear from the context that 2 is the API level and 4 is the number of arguments. I think leaving the explicit numbers and not hidden behind the macros makes the code more readable here. Maybe changing the `int api` to `const int monolog_api` and `size_t argc` to `const size_t add_record_argc` might. Then the code would read:
```
  if (2 <= monolog_api && 4 <= add_record_argc) {
```
What do you think?

Reply:
@mfulb I hope c570e48d08a7ba55fb48c0adbfa44f0403b41066 increases code clarity and maintainability